Modifications to support snap sync

piersy opened the pull request:
This PR contains a collection of additional changes that were required to support snap syncing.

They are:

* Handling celo block receipts - celo-blockchain would add an extra block receipt to the end of the receipts in a block containing any logs generated from system calls executed in that block. So we need to handle that quirk here, in order to support handling migrated legacy data.
* Ensuring non nil difficulty in pre-gingerbread blocks. Before the gingerbread hardfork celo blocks did not have a difficulty field. While the obvious choice to represent this is nil, the go-ethereum codebase assumes non nil difficulties, and so rather than try to catch all locations which  try to access difficulty it is simpler to set a zero valued difficulty on pre gingerbread blocks.
*  Skip the downloader uncle hash check for pre-gingerbread blocks. Pre-gingerbread blocks lacked an uncleHash field (and uncles), but as the field is a non pointer hash in the op-geth structure, when decoding pre-gingerbread blocks the uncleHash is a zero valued hash. go-ethereum expects a block that lacks uncles to have a hash of the empty list which is not a zero valued hash, so in order to allow the downloader to process pre gingerbread blocks we skip this check for pre-gingerbread blocks.
* Handle the base fee calculation for the transition block. The EIP-1559 implementation provides a method to calculate the base fee for a block based on the parent block, this is used in block construction and validation. We set the london hardfork activation block to the the "l2 genesis" and when validating the first EIP-1559 block (I.E the current block is on the london hardfork and the parent is not) the method considers the base fee to be `params.InitialBaseFee` which is a hardcoded value. That doesn't work in our case because we probably want to set the base fee to match that defined at the transition point which can vary, so instead we take the value of the parent block, and if that is missing fall back to the base fee.

piersy committed:
Add code to handle block receipts

This pr adds code to handle the extra block receipt added by celo to the
list of receipts in a block and also adds the GetBlockReceipt rpc api
which was an additional api added by celo to allow retrieveal of the
block receipt.

piersy committed:
Ensure non nil difficulty in pre-gingerbread blocks

Nil difficulty in headers causes problems because it is assumed by the
go-ethereum codebase that difficulty is non nil.

The specific problem requiring this fix was a NPE in
Downloader.processHeaders where the total difficulty (td) is calculated
by adding individual header difficulty values.

Rather than try to fix this specific case it seems safer to ensure
blocks have a non nil Difficulty which should help future proof since it
upholds the assumptions of the go-ethereum codebase.

piersy committed:
Add method to header to check for pre-gingerbread

piersy committed:
Skip uncle hash check for pre gingerbread headers

Pre gingerbread headers do not have a valid uncle hash because they have
no notion of uncles.

piersy committed:
Handle base fee calculation for transition block

palango commented on internal/ethapi/celo_block_receipt.go:
```diff
@@ -0,0 +1,60 @@
+)
+
+// GetBlockReceipt returns "system calls" receipt for the block with the given block hash.
+func (s *BlockChainAPI) GetBlockReceipt(ctx context.Context, hash common.Hash) (map[string]interface{}, error) {
```
Please add a test for this endpoint.

piersy replied:
Added an issue for this here - https://github.com/celo-org/op-geth/issues/156

palango commented on core/types/receipt.go:
```diff
@@ -600,13 +600,16 @@ func (rs Receipts) DeriveFields(config *params.ChainConfig, hash common.Hash, nu
-	if len(txs) != len(rs) {
+
+	// If rs are one longer than txs it indicates the presence of a celo block receipt.
+	if len(txs) != len(rs) && len(txs)+1 != len(rs) {
```
In the old code we also check the hash of the receipt: https://github.com/celo-org/celo-blockchain/blob/e67c1fd73f4e4441bc55a5b3793ba11549a53d2d/core/rawdb/accessors_chain.go#L703

piersy replied:
That code was broken, it was checking the receiptBlock hash matched the receiptTx hash, but those are only set further down in DeriveFields. More detailed description here - https://github.com/celo-org/celo-blockchain/issues/2288

palango commented on core/genesis.go:
```diff
@@ -485,6 +485,9 @@ func (g *Genesis) ToBlock() *types.Block {
 		}
+	} else if g.Difficulty == nil {
+		// In the case of migrated chains we ensure a zero rather than nil difficulty.
+		head.Difficulty = new(big.Int)
```
Mh, after the issue Karl found I wonder if this will bring trouble in the future: https://github.com/celo-org/op-geth/pull/153

With this change, `isMerge` will be true for all blocks. 🤔

piersy replied:
Could you elaborate? The linked issue seems to be about setting the Random field in the block context for vm execution. I don't see how that connects with setting difficulty here?

palango replied:
So if I'm reading the code correctly it will set the difficulty field for the genesis block. This will set the `isMerge` field for that block to true, which is not what we want. (https://github.com/celo-org/op-geth/blob/celo6/params/config.go#L1097)

However, that might not cause any problems. I just wanted to point it out because I was a bit surprised to see this behaviour in PR from Karl.

piersy replied:
Ok got it, in `NewEvmBlockContext` `Random` is set to the header's mix hash if difficulty is equal to zero. It seems like this shouldn't affect us since we are not supporting execution for historical blocks, so we shouldn't end up in this code path. If we didn't set this to zero then we'd run into a nil pointer exception.

palango commented on consensus/misc/eip1559/eip1559.go:
```diff
@@ -58,6 +58,12 @@ func VerifyEIP1559Header(config *params.ChainConfig, parent, header *types.Heade
+	// If this is the cel2 transition block then we use the initial base fee
+	// from the Cel2Config of the chain config.
+	if config.Cel2Time != nil && *config.Cel2Time == time {
+		return new(big.Int).SetUint64(config.Cel2Config.TransitionBlockBaseFee)
```
Do we need this to be configurable? Can't we just hardcode it (there will be no txs in the block anyways) or use the basefee of the block before?

piersy replied:
Hardcoding is what ethereum did with `InitialBaseFee` but we probably don't want to change that, and it doesn't match up with our current base fees. Just taking the base fee from the block works for migrating post gingerbread blocks, but not for pre gingerbread. So we could do away with this, but we'd lose the ability to migrate pre-gingerbread blocks.

palango replied:
We'll probably never migrate pre-gingerbread blocks, so I'd avoid the complexity. But no hard opinion on my side.

piersy replied:
It's useful if you want to test locally with a small portion of one of our established networks, but that's it.

So I think I'm going to go for just taking the parent base fee value, and I'll update the migration script to fail if an attempt is made to migrate at a pre-gingerbread block.

palango replied:
Or just use some reasonable default in that case?

piersy replied:
Yeah also could do that.

piersy replied:
Hi @palango updated the approach to choosing the base fee here.

piersy committed:
Merge branch 'celo6' into piersy/modifications-to-support-snap-sync

piersy committed:
Revert "Handle base fee calculation for transition block"

This reverts commit 9795acac4fa262334dd1114b7997f84b533bdcc5.

piersy committed:
Change approach to choosing transition block base fee

If the parent block has a base fee then we use that, otherwise we use
the initial base fee.

piersy committed:
Merge branch 'celo6' into piersy/modifications-to-support-snap-sync
